Compilation error with gcc

Hi,

first of all thanks for all your work, it's an amazing project that helps me a lot. I just downloaded the recent version of xtensor and xtl and get the following compiler error:

In file included from /home/philipp/eclipse/mlhpn/xtensor/include/xtensor/xsemantic.hpp:16:0,
                 from /home/philipp/eclipse/mlhpn/xtensor/include/xtensor/xstrided_view.hpp:25,
                 from /home/philipp/eclipse/mlhpn/xtensor/include/xtensor/xoperation.hpp:22,
                 from /home/philipp/eclipse/mlhpn/xtensor/include/xtensor/xmath.hpp:27,
                 from /home/philipp/eclipse/mlhpn/xtensor/include/xtensor/xcontainer.hpp:24,
                 from /home/philipp/eclipse/mlhpn/xtensor/include/xtensor/xarray.hpp:20,
                 from /home/philipp/git/mlhpn/libraries/fekernel/inc/alias.hpp:4,
                 from /home/philipp/git/mlhpn/libraries/fekernel/inc/spacetree.hpp:4,
                 from /home/philipp/git/mlhpn/libraries/fekernel/src/spacetree.cpp:3:
/home/philipp/eclipse/mlhpn/xtensor/include/xtensor/xassign.hpp: In instantiation of ‘static void xt::strided_loop_assigner<simd>::run(E1&, const E2&) [with E1 = xt::xstrided_view<xt::xarray_container<xt::uvector<bool, std::allocator<bool> >, (xt::layout_type)1, xt::svector<long unsigned int, 4, std::allocator<long unsigned int>, true>, xt::xtensor_expression_tag>&, xt::svector<long unsigned int, 4, std::allocator<long unsigned int>, true>, (xt::layout_type)0, xt::detail::inner_storage_getter<xt::xarray_container<xt::uvector<bool, std::allocator<bool> >, (xt::layout_type)1, xt::svector<long unsigned int, 4, std::allocator<long unsigned int>, true>, xt::xtensor_expression_tag>&> >; E2 = xt::xarray_container<xt::uvector<bool, std::allocator<bool> >, (xt::layout_type)0, xt::svector<long unsigned int, 4, std::allocator<long unsigned int>, true>, xt::xtensor_expression_tag>; bool simd = true]’:
/home/philipp/eclipse/mlhpn/xtensor/include/xtensor/xstrided_view.hpp:353:45:   required from ‘void xt::xstrided_view_detail::run_assign_temporary_impl(V&, const T&, std::true_type) [with V = xt::xstrided_view<xt::xarray_container<xt::uvector<bool, std::allocator<bool> >, (xt::layout_type)1, xt::svector<long unsigned int, 4, std::allocator<long unsigned int>, true>, xt::xtensor_expression_tag>&, xt::svector<long unsigned int, 4, std::allocator<long unsigned int>, true>, (xt::layout_type)0, xt::detail::inner_storage_getter<xt::xarray_container<xt::uvector<bool, std::allocator<bool> >, (xt::layout_type)1, xt::svector<long unsigned int, 4, std::allocator<long unsigned int>, true>, xt::xtensor_expression_tag>&> >; T = xt::xarray_container<xt::uvector<bool, std::allocator<bool> >, (xt::layout_type)0, xt::svector<long unsigned int, 4, std::allocator<long unsigned int>, true>, xt::xtensor_expression_tag>; std::true_type = std::integral_constant<bool, true>]’
/home/philipp/eclipse/mlhpn/xtensor/include/xtensor/xstrided_view.hpp:367:56:   required from ‘void xt::xstrided_view<CT, S, L, FST>::assign_temporary_impl(xt::xstrided_view<CT, S, L, FST>::temporary_type&&) [with CT = xt::xarray_container<xt::uvector<bool, std::allocator<bool> >, (xt::layout_type)1, xt::svector<long unsigned int, 4, std::allocator<long unsigned int>, true>, xt::xtensor_expression_tag>&; S = xt::svector<long unsigned int, 4, std::allocator<long unsigned int>, true>; xt::layout_type L = (xt::layout_type)0; FST = xt::detail::inner_storage_getter<xt::xarray_container<xt::uvector<bool, std::allocator<bool> >, (xt::layout_type)1, xt::svector<long unsigned int, 4, std::allocator<long unsigned int>, true>, xt::xtensor_expression_tag>&>; xt::xstrided_view<CT, S, L, FST>::temporary_type = xt::xarray_container<xt::uvector<bool, std::allocator<bool> >, (xt::layout_type)0, xt::svector<long unsigned int, 4, std::allocator<long unsigned int>, true>, xt::xtensor_expression_tag>]’
/home/philipp/eclipse/mlhpn/xtensor/include/xtensor/xsemantic.hpp:662:9:   required from ‘xt::xview_semantic<D>::derived_type& xt::xview_semantic<D>::assign_temporary(xt::xview_semantic<D>::temporary_type&&) [with D = xt::xstrided_view<xt::xarray_container<xt::uvector<bool, std::allocator<bool> >, (xt::layout_type)1, xt::svector<long unsigned int, 4, std::allocator<long unsigned int>, true>, xt::xtensor_expression_tag>&, xt::svector<long unsigned int, 4, std::allocator<long unsigned int>, true>, (xt::layout_type)0, xt::detail::inner_storage_getter<xt::xarray_container<xt::uvector<bool, std::allocator<bool> >, (xt::layout_type)1, xt::svector<long unsigned int, 4, std::allocator<long unsigned int>, true>, xt::xtensor_expression_tag>&> >; xt::xview_semantic<D>::derived_type = xt::xstrided_view<xt::xarray_container<xt::uvector<bool, std::allocator<bool> >, (xt::layout_type)1, xt::svector<long unsigned int, 4, std::allocator<long unsigned int>, true>, xt::xtensor_expression_tag>&, xt::svector<long unsigned int, 4, std::allocator<long unsigned int>, true>, (xt::layout_type)0, xt::detail::inner_storage_getter<xt::xarray_container<xt::uvector<bool, std::allocator<bool> >, (xt::layout_type)1, xt::svector<long unsigned int, 4, std::allocator<long unsigned int>, true>, xt::xtensor_expression_tag>&> >; xt::xview_semantic<D>::temporary_type = xt::xarray_container<xt::uvector<bool, std::allocator<bool> >, (xt::layout_type)0, xt::svector<long unsigned int, 4, std::allocator<long unsigned int>, true>, xt::xtensor_expression_tag>]’
/home/philipp/eclipse/mlhpn/xtensor/include/xtensor/xstrided_view.hpp:322:53:   required from ‘xt::xstrided_view<CT, S, L, FST>& xt::xstrided_view<CT, S, L, FST>::operator=(const xt::xstrided_view<CT, S, L, FST>&) [with CT = xt::xarray_container<xt::uvector<bool, std::allocator<bool> >, (xt::layout_type)1, xt::svector<long unsigned int, 4, std::allocator<long unsigned int>, true>, xt::xtensor_expression_tag>&; S = xt::svector<long unsigned int, 4, std::allocator<long unsigned int>, true>; xt::layout_type L = (xt::layout_type)0; FST = xt::detail::inner_storage_getter<xt::xarray_container<xt::uvector<bool, std::allocator<bool> >, (xt::layout_type)1, xt::svector<long unsigned int, 4, std::allocator<long unsigned int>, true>, xt::xtensor_expression_tag>&>]’
/home/philipp/git/mlhpn/libraries/fekernel/src/spacetree.cpp:45:21:   recursively required from ‘void mlhpn::fekernel::expandDimensions(size_t, size_t, xt::xarray<ElementType>&, Args& ...) [with ElementType = unsigned char; Args = {xt::xarray_container<xt::uvector<bool, std::allocator<bool> >, (xt::layout_type)1, xt::svector<long unsigned int, 4, std::allocator<long unsigned int>, true>, xt::xtensor_expression_tag>, xt::xarray_container<xt::uvector<unsigned int, std::allocator<unsigned int> >, (xt::layout_type)1, xt::svector<long unsigned int, 4, std::allocator<long unsigned int>, true>, xt::xtensor_expression_tag>, xt::xarray_container<xt::uvector<long unsigned int, std::allocator<long unsigned int> >, (xt::layout_type)1, xt::svector<long unsigned int, 4, std::allocator<long unsigned int>, true>, xt::xtensor_expression_tag>}; size_t = long unsigned int; xt::xarray<ElementType> = xt::xarray_container<xt::uvector<unsigned char, std::allocator<unsigned char> >, (xt::layout_type)1, xt::svector<long unsigned int, 4, std::allocator<long unsigned int>, true>, xt::xtensor_expression_tag>; typename std::vector<X>::size_type = long unsigned int; typename xt::uvector<T, std::allocator<_CharT> >::size_type = long unsigned int]’
/home/philipp/git/mlhpn/libraries/fekernel/src/spacetree.cpp:45:21:   required from ‘void mlhpn::fekernel::expandDimensions(size_t, size_t, xt::xarray<ElementType>&, Args& ...) [with ElementType = long unsigned int; Args = {xt::xarray_container<xt::uvector<unsigned char, std::allocator<unsigned char> >, (xt::layout_type)1, xt::svector<long unsigned int, 4, std::allocator<long unsigned int>, true>, xt::xtensor_expression_tag>, xt::xarray_container<xt::uvector<bool, std::allocator<bool> >, (xt::layout_type)1, xt::svector<long unsigned int, 4, std::allocator<long unsigned int>, true>, xt::xtensor_expression_tag>, xt::xarray_container<xt::uvector<unsigned int, std::allocator<unsigned int> >, (xt::layout_type)1, xt::svector<long unsigned int, 4, std::allocator<long unsigned int>, true>, xt::xtensor_expression_tag>, xt::xarray_container<xt::uvector<long unsigned int, std::allocator<long unsigned int> >, (xt::layout_type)1, xt::svector<long unsigned int, 4, std::allocator<long unsigned int>, true>, xt::xtensor_expression_tag>}; size_t = long unsigned int; xt::xarray<ElementType> = xt::xarray_container<xt::uvector<long unsigned int, std::allocator<long unsigned int> >, (xt::layout_type)1, xt::svector<long unsigned int, 4, std::allocator<long unsigned int>, true>, xt::xtensor_expression_tag>; typename std::vector<X>::size_type = long unsigned int; typename xt::uvector<T, std::allocator<_CharT> >::size_type = long unsigned int]’
/home/philipp/git/mlhpn/libraries/fekernel/src/spacetree.cpp:202:25:   required from ‘void mlhpn::fekernel::MultilevelGrid<D>::refine(mlhpn::fekernel::RefinementFunction<D>&) [with long unsigned int D = 2; mlhpn::fekernel::RefinementFunction<D> = std::function<bool(const mlhpn::fekernel::AbsCell<2>&)>]’
/home/philipp/git/mlhpn/libraries/fekernel/src/spacetree.cpp:383:16:   required from here
/home/philipp/eclipse/mlhpn/xtensor/include/xtensor/xassign.hpp:860:49: error: ‘size’ is not a member of ‘simd_type {aka bool}’
         std::size_t simd_size = inner_loop_size / simd_type::size;
                                 ~~~~~~~~~~~~~~~~^~~~~~~~~~~
/home/philipp/eclipse/mlhpn/xtensor/include/xtensor/xassign.hpp:861:49: error: ‘size’ is not a member of ‘simd_type {aka bool}’
         std::size_t simd_rest = inner_loop_size % simd_type::size;
                                 ~~~~~~~~~~~~~~~~^~~~~~~~~~~
/home/philipp/eclipse/mlhpn/xtensor/include/xtensor/xassign.hpp:882:77: error: use of ‘auto xt::conditional_cast(U&&) [with bool condition = false; T = bool; U = const bool&]’ before deduction of ‘auto’
                 *(res_stepper) = conditional_cast<needs_cast, e1_value_type>(*(fct_stepper));

---------------------------------

I went back to the old version and did not have any issues.

Thanks again!

Philipp

--- a/include/xtensor/xassign.hpp
+++ b/include/xtensor/xassign.hpp
@@ -298,8 +298,10 @@ namespace xt
         static constexpr bool contiguous_layout() { return E1::contiguous_layout && E2::contiguous_layout; }
         static constexpr bool convertible_types() { return std::is_convertible<e2_value_type, e1_value_type>::value; }
 
+        static constexpr bool use_xsimd() { return xt_simd::simd_traits<int8_t>::size > 1; }
+
         template <class T>
-        static constexpr bool simd_size_impl() { return xt_simd::simd_traits<T>::size > 1 || is_bool<T>::value; }
+        static constexpr bool simd_size_impl() { return xt_simd::simd_traits<T>::size > 1 || (is_bool<T>::value && use_xsimd()); }
         static constexpr bool simd_size() { return simd_size_impl<e1_value_type>() && simd_size_impl<e2_value_type>(); }
         static constexpr bool simd_interface() { return has_simd_interface<E2, requested_value_type>(); }
         static constexpr bool simd_assign() { return convertible_types() && simd_size() && simd_interface(); }

--- a/include/xtensor/xtensor_simd.hpp
+++ b/include/xtensor/xtensor_simd.hpp
@@ -299,7 +299,7 @@ namespace xt
 
     template <class C, class T1, class T2>
     struct container_simd_return_type
-        : std::enable_if<!forbid_simd<C>::value, xt_simd::simd_return_type<T1, T2>>
+        : std::enable_if<!forbid_simd<C>::value, xt_simd::simd_return_type<T1, bool_load_type<T2>>>
     {
     };
 

--- a/test/CMakeLists.txt
+++ b/test/CMakeLists.txt
@@ -54,7 +54,7 @@ endif()
 
 if(CMAKE_CXX_COMPILER_ID MATCHES "GNU" OR (CMAKE_CXX_COMPILER_ID MATCHES "Intel" AND NOT WIN32))
   set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} ${_cxx_std_flag} -march=native -Wunused-parameter -Wextra -Wreorder -Wconversion -Wsign-conversion")
-  set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -Wold-style-cast -Wunused-variable")
+  set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -Wold-style-cast -Wunused-variable -ftemplate-backtrace-limit=0")
 elseif(CMAKE_CXX_COMPILER_ID MATCHES "MSVC")
   set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} ${_cxx_std_flag} /EHsc /MP /bigobj")
   set(CMAKE_EXE_LINKER_FLAGS ${CMAKE_EXE_LINKER_FLAGS} /MANIFEST:NO)

--- a/test/test_xview.cpp
+++ b/test/test_xview.cpp
@@ -1477,4 +1477,15 @@ namespace xt
         noalias(view(reshape_view(a, {3}), xt::all())) = b;
         EXPECT_EQ(tmp, res);
     }
+
+    TEST(xview, view_on_bool)
+    {
+        xt::xarray<bool> a { { false, false }, { false, false } };
+        xt::xarray<bool> b { {  true,  true }, {  true,  true } };
+        xt::view( a, 0 ) = xt::view( b, 0 );
+        EXPECT_TRUE(a(0, 0));
+        EXPECT_TRUE(a(0, 1));
+        EXPECT_FALSE(a(1, 0));
+        EXPECT_FALSE(a(1, 1));
+    }
 }

